### PR TITLE
Added new package provider for foreign exchange rates.

### PR DESCRIPTION
**Pull request by shazbert**
- Contains segregated provider packages
- Waiting for review before full integration

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -99,6 +99,7 @@ type Config struct {
 	SMS                      SMSGlobalConfig  `json:"SMSGlobal"`
 	Webserver                WebserverConfig  `json:"Webserver"`
 	Exchanges                []ExchangeConfig `json:"Exchanges"`
+	Providers                []ProviderConfig `json:"Providers"`
```
Comment (thrasher-) [inline config/config.go]: Need to come up with a more specific name, perhaps ForexProviders or ForexConversionProviders.

File: config/config.go [hunk 2]
```diff
@@ -230,6 +241,20 @@ func (c *Config) GetExchangeConfig(name string) (ExchangeConfig, error) {
 	return ExchangeConfig{}, fmt.Errorf(ErrExchangeNotFound, name)
 }
 
+// GetProviderConfig returns a provider configuration by its name
+func (c *Config) GetProviderConfig(name string) (ProviderConfig, error) {
+	c.m.Lock()
+	defer c.m.Unlock()
+	log.Println("What", len(c.Providers))
+	for i := range c.Providers {
+		log.Println(c.Providers[i].Name)
```
Comment (thrasher-) [inline config/config.go]: Just need to make sure this is removed upon merge.

File: config/config.go [hunk 3]
```diff
@@ -341,6 +366,20 @@ func (c *Config) CheckWebserverConfigValues() error {
 	return nil
 }
 
+// CheckProviderConfigValues checks Provider configuration values
+func (c *Config) CheckProviderConfigValues() error {
```
Comment (thrasher-) [inline config/config.go]: A check here for default values would be good. Fatal error if the default values are set without the users ones unless the forex provider allows forex conversions without an API key.

File: config/config.go (hunk 1 above)
Comment (shazbert) [reply config/config.go]: ForexRateProviders?

File: config/config.go [hunk 4]
```diff
@@ -233,6 +245,20 @@ func (c *Config) GetExchangeConfig(name string) (ExchangeConfig, error) {
 	return ExchangeConfig{}, fmt.Errorf(ErrExchangeNotFound, name)
 }
 
+// GetProviderConfig returns a forex provider configuration by its name
+func (c *Config) GetProviderConfig(name string) (ForexProviderConfig, error) {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	for i := range c.ForexProviders {
+		log.Println(c.ForexProviders[i].Name)
```
Comment (thrasher-) [inline config/config.go]: Can remove this Println

File: config/config.go [hunk 5]
```diff
@@ -97,10 +97,11 @@ type Config struct {
...
 	GlobalHTTPTimeout        time.Duration
-	Portfolio                portfolio.Base   `json:"PortfolioAddresses"`
-	SMS                      SMSGlobalConfig  `json:"SMSGlobal"`
-	Webserver                WebserverConfig  `json:"Webserver"`
-	Exchanges                []ExchangeConfig `json:"Exchanges"`
+	Portfolio                portfolio.Base        `json:"PortfolioAddresses"`
+	SMS                      SMSGlobalConfig       `json:"SMSGlobal"`
+	Webserver                WebserverConfig       `json:"Webserver"`
+	Exchanges                []ExchangeConfig      `json:"Exchanges"`
+	ForexProviders           []ForexProviderConfig `json:"Providers"`
```
Comment (thrasher-) [inline config/config.go]: Should make the Providers JSON ouput match the ForexProviders var

File: config/config.go [hunk 6]
```diff
@@ -349,6 +375,23 @@ func (c *Config) CheckWebserverConfigValues() error {
...
+	for i := range c.ForexProviders {
+		if c.ForexProviders[i].Enabled == true {
+			if c.ForexProviders[i].APIKey == "Key" {
+				log.Fatal("provider api key not set - please set via your config.json file")
+			}
+			count++
+		}
+	}
+	if count < 1 {
+		log.Println("warning config.go CheckProviderConfigValues() no providers set")
```
Comment (thrasher-) [inline config/config.go]: Moving forward we will need to fatal error if there's no forex providers set (After the free unauthenticated fixer.io service is stopped).

Comment (shazbert) [reply config/config.go]: Added to trello.

**Comment by thrasher-** [comment]
Squashed and placed in https://github.com/thrasher-/gocryptotrader/commit/58051b89c71737daa47fde09f968c65762977021

**Pull request closed without merging**